statistics/semantic/src/main/java/com/spotify/heroic/statistics/semantic/SemanticMetricBackendReporter.java

```diff
@@ -147,16 +175,35 @@ public void reportDataNoLongerNeeded(final long n) {
             public void reportOperationEnded() {
                 // Decrement any remaining live memory
                 sampleSizeLive.dec(dataInMemory.get());
+
+                final long dataReadLocal = dataRead.get();
                 /* The accumulated count has previously been incremented with all of the data read,
                  * so decrease the full amount now */
-                sampleSizeAccumulated.dec(dataRead.get());
+                sampleSizeAccumulated.dec(dataReadLocal);
 
-                querySamplesRead.update(dataRead.get());
+                querySamplesRead.update(dataReadLocal);
                 queryRowsAccessed.update(rowsAccessed.get());
+                queryMaxLiveSamples.update(maxDataInMemory.get());
+
+                long duration = queryWatch.elapsed(TimeUnit.NANOSECONDS);
+                if (duration != 0) {
+                    queryReadRate.update((1000_000_000 * dataReadLocal) / duration);
+                }
+
+                for (ReadRangeStatistics stats : seriesDensity.values()) {
+                    long rangeMs = stats.getRange().diff();
+                    queryRowDensity.update((3600_000L * stats.getValue().get()) / rangeMs);
```

Comment:
Guessing here, but but a per-second value would probably round to zero in most cases. Most of our applications emits data around every 30 seconds IIRC.

Flipping the units might be prudent in this case (ms / sample). The highest theoretical density we support is 1ms, and we currently don't support non-long histograms.